File: ui/src/wire.js
/* J1 wire layer — JavaScript port of engine/wire.py (PFQ v1 / PFR v1).
 * Byte-exact agreement enforced by ui/tests/diff-harness.mjs.
 */
import {
  sweep, validateRoom, roomClasses, validateNonce, parseDid,
  canonicalMsg, canonicalNote, makeVerifier, estimateRequestLine,
  MAX_TEXT_CHARS, PreflightError, pyRepr, ok, bad,
} from "./engine.js";

// Python tuple rendering for error details (str(('a', 'b', 'c'))).
const pyTupleStr = (xs) => "(" + xs.map(x => `'${x}'`).join(", ") + ")";

// ---------------------------------------------------------------- constants
export const PFQ_PREFIX = "PFQ v1";
export const PFR_PREFIX = "PFR v1";
export const OPS = ["preview", "verify", "audit-did-note"];
export const STATUSES = ["PASS", "FAIL", "PARTIAL", "ERROR"];

export const REJECT_CODES = new Set([
  "E_EMPTY_AFTER_SWEEP", "E_TEXT_TOO_LONG", "E_BAD_ROOM", "E_BAD_NONCE_FORMAT",
  "E_BAD_DID", "E_BAD_SIG_ENCODING", "E_SIG_INVALID", "E_CANONICAL_TOO_LONG",
]);
export const WARN_CODES = new Set([
  "W_SWEPT_CHARS", "W_URL_LONG", "W_LEADING_ZERO_NONCE",
  "W_NOTE_WRONG_KEY", "W_NOTE_LEGACY_PATH", "W_NOTE_FIELD_MISMATCH",
]);
export const OBSERVE_CODES = new Set([
  "O_NONCE_FLOOR_VISIBLE", "O_NO_PRIOR_WRITES_SEEN",
  "O_ROOM_OWNED", "O_CAPACITY_TIGHT",
]);

export const CID_RE = /^[0-9a-f]{16}$/;
const SEMVER_RE = /^\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?$/;
const KEY_RE = /^[a-z][a-z0-9_]{0,23}$/;
const SHA256_RE = /^[0-9a-f]{64}$/;
const FP_RE = /^[0-9a-f]{16}$/;
const NS_RE = /^did(?:-[0-9a-f]{2})?$/;
const NOTE_KEY_RE = /^[a-z0-9][a-z0-9_-]{0,47}$/;
const REF_RE = /^[0-9A-Za-z._:@/-]{1,40}$/;

export class WireError extends Error {
  constructor(code, detail = "") {
    super(detail ? `${code}: ${detail}` : code);
    this.name = "WireError";
    this.code = code;
    this.detail = detail;
  }
}

// ----------------------------------------------------------- value escaping
export function encodeValue(value) {
  let out = "";
  for (const ch of value) {
    if (ch === "%") out += "%25";
    else if (ch === "|") out += "%7C";
    else if (ch === ";") out += "%3B";
    else out += ch;
  }
  return out;
}

export function decodeValue(value) {
  let out = "";
  for (let i = 0; i < value.length; i++) {
    const ch = value[i];
    if (ch !== "%") { out += ch; continue; }
    const pair = value.slice(i + 1, i + 3);
    if (pair === "25") { out += "%"; i += 2; }
    else if (pair === "7C") { out += "|"; i += 2; }
    else if (pair === "3B") { out += ";"; i += 2; }
    else throw new WireError("X_BAD_ENCODING",
      `only %25 %7C %3B escapes exist; found %${pyRepr(pair)} at offset ${i}`);
  }
  return out;
}

// -------------------------------------------------------------- shared bits
function checkLine(line) {
  if (typeof line !== "string") throw new WireError("X_BAD_TYPE", "wire line must be str");
  if ([...line].length > MAX_TEXT_CHARS) {
    throw new WireError("X_LINE_TOO_LONG",
      `${[...line].length} chars exceeds message cap ${MAX_TEXT_CHARS}`);
  }
  try {
    const swept = sweep(line);
    if (swept.stored !== line) {
      throw new WireError("X_NOT_SWEEP_SAFE",
        "line contains characters the server would replace or strip; send the post-sweep form");
    }
  } catch (e) {
    if (e instanceof WireError) throw e;
    throw new WireError("X_EMPTY_AFTER_SWEEP", "line has no visible content");
  }
}

function parseCid(field) {
  const cid = field.trim();
  if (!CID_RE.test(cid)) {
    throw new WireError("X_BAD_CID",
      `cid must be exactly 16 lowercase hex chars, got ${pyRepr(field)}`);
  }
  return cid;
}

function splitParams(raw) {
  const pairs = [];
  for (let tok of raw.split(";")) {
    tok = tok.trim();
    if (!tok) throw new WireError("X_BAD_PARAM", "empty parameter token");
    const eq = tok.indexOf("=");
    if (eq <= 0) throw new WireError("X_BAD_PARAM", `parameter ${pyRepr(tok)} lacks 'key=value' form`);
    const key = tok.slice(0, eq).trim();
    if (!KEY_RE.test(key)) throw new WireError("X_BAD_KEY",
      `parameter key ${pyRepr(key)} must match [a-z][a-z0-9_]{0,23}`);
    const val = tok.slice(eq + 1).trim();
    if (!val) throw new WireError("X_EMPTY_VALUE", `parameter ${pyRepr(key)} has an empty value`);
    pairs.push([key, decodeValue(val)]);
  }
  return pairs;
}

function dupCheck(pairs) {
  const seen = {};
  for (const [k] of pairs) {
    if (k in seen) throw new WireError("X_DUPLICATE_KEY", `parameter ${pyRepr(k)} appears twice`);
    seen[k] = true;
  }
}

function requireFirst(pairs, key) {
  if (!pairs.length || pairs[0][0] !== key)
    throw new WireError("X_ORDER", `first parameter must be ${key}=…`);
}

export function validateReplyRoom(room) {
  const v = validateRoom(room);
  if (!v.ok) throw new WireError("X_BAD_REPLY_ROOM", v.detail || "invalid room");
  if (!roomClasses(room).includes("mb"))
    throw new WireError("X_BAD_REPLY_ROOM", `${pyRepr(room)} lacks the mb- class; replies go to mailboxes`);
}

// ------------------------------------------------------------- op schemas
export const OP_SCHEMAS = {
  preview: { required: ["room", "nonce", "did", "text"], optional: ["sig"] },
  verify: { required: ["did"], optional: ["sig", "nonce", "room", "text", "canonical", "sha256"] },
  "audit-did-note": { required: ["value"], optional: ["did", "fp", "ns", "key"] },
};

function schemaCheck(op, kv) {
  const { required, optional } = OP_SCHEMAS[op];
  const unknown = Object.keys(kv).filter(k => k !== "reply" && !required.includes(k) && !optional.includes(k));
  if (unknown.length) throw new WireError("X_UNKNOWN_KEY", `op '${op}' does not accept ${unknown.sort()}`);
  const missing = required.filter(k => !(k in kv));
  if (missing.length) throw new WireError("X_MISSING_KEY", `op '${op}' requires ${missing}`);
  if ("sha256" in kv && !SHA256_RE.test(kv.sha256))
    throw new WireError("X_BAD_SHA256", "must be 64 lowercase hex chars");
  if ("fp" in kv && !FP_RE.test(kv.fp))
    throw new WireError("X_BAD_FP", "must be 16 lowercase hex chars");
  if ("ns" in kv && !NS_RE.test(kv.ns))
    throw new WireError("X_BAD_NS", "expected 'did' or 'did-<2 hex>'");
  if ("key" in kv && !NOTE_KEY_RE.test(kv.key))
    throw new WireError("X_BAD_NOTE_KEY", "invalid note key name");

  if (op === "verify") {
    const full = ["nonce", "room", "text"].map(k => k in kv);
    const privacy = ["canonical", "sha256"].map(k => k in kv);
    if (full.some(Boolean) && !full.every(Boolean))
      throw new WireError("X_AMBIGUOUS_MODE", "full mode needs nonce, room AND text together");
    if (full.some(Boolean) && privacy.some(Boolean))
      throw new WireError("X_AMBIGUOUS_MODE", "full mode and privacy mode are mutually exclusive");
    if (!full.some(Boolean)) {
      if (!privacy.some(Boolean))
        throw new WireError("X_MISSING_KEY", "verify needs nonce+room+text or canonical+sha256");
      if (!privacy.every(Boolean))
        throw new WireError("X_AMBIGUOUS_MODE", "privacy mode needs canonical AND sha256 together");
    }
  }
  if (op === "audit-did-note") {
    const hasDid = "did" in kv, hasFp = "fp" in kv;
    if (hasDid && hasFp) throw new WireError("X_AMBIGUOUS_MODE", "give exactly one of did= or fp=");
    if (!hasDid && !hasFp) throw new WireError("X_MISSING_KEY", "audit-did-note needs did= or fp=");
  }
}

// ------------------------------------------------------------------- PFQ
export function parsePfq(line) {
  checkLine(line);
  const parts = line.split("|");
  if (parts.length !== 4) {
    throw new WireError("X_BAD_STRUCTURE",
      `expected exactly 3 unescaped '|' separators, got ${parts.length - 1}`);
  }
  const [prefixF, cidF, opF, paramsF] = parts.map(p => p.trim());
  if (prefixF !== PFQ_PREFIX) throw new WireError("X_BAD_PREFIX", `expected ${pyRepr(PFQ_PREFIX)}, got ${pyRepr(prefixF)}`);
  const cid = parseCid(cidF);
  const op = opF;
  if (!OPS.includes(op)) throw new WireError("X_BAD_OP", `op ${pyRepr(op)} outside ${pyTupleStr(OPS)}`);
  const pairs = splitParams(paramsF);
  requireFirst(pairs, "reply");
  dupCheck(pairs);
  const kv = Object.fromEntries(pairs);
  validateReplyRoom(kv.reply);
  schemaCheck(op, kv);
  return { kind: "PFQ", cid, op, params: kv };
}

export function renderPfq(q) {
  const op = q.op;
  if (!OPS.includes(op)) throw new WireError("X_BAD_OP", `op ${JSON.stringify(op)} outside ['preview','verify','audit-did-note']`);
  const cid = q.cid ?? "";
  if (!CID_RE.test(cid)) throw new WireError("X_BAD_CID", "cid must be 16 lowercase hex chars");
  const kv = q.params ?? {};
  if (Object.keys(kv).length !== new Set(Object.keys(kv)).size)
    throw new WireError("X_DUPLICATE_KEY", "dict cannot hold duplicate keys");
  if (!("reply" in kv)) throw new WireError("X_MISSING_KEY", "reply= is mandatory");
  schemaCheck(op, kv);
  const ordered = ["reply", ...Object.keys(kv).filter(k => k !== "reply")];
  const params = ordered.map(k => `${k}=${encodeValue(kv[k])}`).join(" ; ");
  return `${PFQ_PREFIX} | ${cid} | ${op} | ${params}`;
}

// -------------------------------------------------------------- findings
function findingToken(f) {
  const [kind, code, ref, detail] = f;
  const enc = encodeValue(detail);
  if (kind === "T1-ok") return "T1-ok" + (enc ? ` ${enc}` : "");
  if (kind === "T1-reject" || kind === "T1-warn")
    return `${kind}:${code}` + (enc ? ` ${enc}` : "");
  if (kind === "T2-observe") {
    const base = `T2-observe:${code}@${ref}`;
    return base + (enc ? ` ${enc}` : "");
  }
  throw new WireError("X_BAD_FINDING", `unrenderable finding kind ${kind}`);
}

// ------------------------------------------------------------------- PFR
export function parsePfr(line) {
  checkLine(line);
  const parts = line.split("|");
  if (parts.length !== 4) {
    throw new WireError("X_BAD_STRUCTURE",
      `expected exactly 3 unescaped '|' separators, got ${parts.length - 1}`);
  }
  const [prefixF, cidF, statusF, paramsF] = parts.map(p => p.trim());
  if (prefixF !== PFR_PREFIX) throw new WireError("X_BAD_PREFIX", `expected ${pyRepr(PFR_PREFIX)}, got ${pyRepr(prefixF)}`);
  const cid = parseCid(cidF);
  const status = statusF;
  if (!STATUSES.includes(status)) throw new WireError("X_BAD_STATUS", `status ${pyRepr(status)} outside ${pyTupleStr(STATUSES)}`);

  const tokens = paramsF.split(";").map(t => t.trim());
  if (!tokens.length || !tokens[0]) throw new WireError("X_BAD_PARAM", "empty parameter section");
  // First token: engine=<semver>. Findings are bare tokens with spaces.
  const eq = tokens[0].indexOf("=");
  if (eq <= 0 || tokens[0].slice(0, eq).trim() !== "engine")
    throw new WireError("X_ORDER", "first parameter must be engine=<semver>");
  const engine = decodeValue(tokens[0].slice(eq + 1).trim());
  if (!engine) throw new WireError("X_EMPTY_VALUE", "engine has an empty value");
  if (!SEMVER_RE.test(engine)) throw new WireError("X_BAD_SEMVER", `engine version ${engine} is not semver`);

  if (status === "ERROR") {
    if (tokens.length !== 2) throw new WireError("X_BAD_STRUCTURE", "ERROR response carries exactly one error= token");
    const eq2 = tokens[1].indexOf("=");
    if (eq2 <= 0 || tokens[1].slice(0, eq2).trim() !== "error")
      throw new WireError("X_BAD_PARAM", "ERROR response requires error=<text>");
    const errText = decodeValue(tokens[1].slice(eq2 + 1).trim());
    if (!errText) throw new WireError("X_EMPTY_VALUE", "error= text is empty");
    return { kind: "PFR", cid, status, engine, findings: [], error: errText };
  }

  const findings = [];
  for (let j = 1; j < tokens.length; j++) {
    const tok = tokens[j];
    if (!tok) throw new WireError("X_BAD_PARAM", "empty parameter token");
    const sp = tok.indexOf(" ");
    const head = sp < 0 ? tok : tok.slice(0, sp);
    const detail = sp < 0 ? "" : decodeValue(tok.slice(sp + 1).trim());
    if (head === "T1-ok") { findings.push(["T1-ok", "", "", detail]); continue; }
    const m1 = head.match(/^(T1-reject|T1-warn):([A-Z0-9_]+)$/);
    if (m1) {
      const frozen = m1[1] === "T1-reject" ? REJECT_CODES : WARN_CODES;
      if (!frozen.has(m1[2])) throw new WireError("X_UNKNOWN_FINDING_CODE",
        `${m1[1]}:${m1[2]} is outside the frozen ${m1[1]} vocabulary`);
      findings.push([m1[1], m1[2], "", detail]);
      continue;
    }
    const m2 = head.match(/^T2-observe:([A-Z0-9_]+)@(.*)$/s);
    if (m2) {
      const [, code, ref] = m2;
      if (!OBSERVE_CODES.has(code)) throw new WireError("X_UNKNOWN_FINDING_CODE",
        `T2-observe:${code} is outside the frozen T2 vocabulary`);
      if (!ref || !REF_RE.test(ref)) throw new WireError("X_BAD_OBSERVATION_REF", `observation ref ${pyRepr(ref)} is malformed`);
      findings.push(["T2-observe", code, ref, detail]);
      continue;
    }
    throw new WireError("X_BAD_FINDING", `finding token ${pyRepr(head)} is malformed`);
  }
  if (!findings.length) throw new WireError("X_NO_FINDINGS", "a PFR carries at least one finding token");
  return { kind: "PFR", cid, status, engine, findings };
}

export function renderPfr(r) {
  const status = r.status;
  if (!STATUSES.includes(status)) throw new WireError("X_BAD_STATUS", `status ${status} outside ${JSON.stringify(STATUSES)}`);
  const cid = r.cid ?? "";
  if (!CID_RE.test(cid)) throw new WireError("X_BAD_CID", "cid must be 16 lowercase hex chars");
  const engine = r.engine ?? "";
  if (!SEMVER_RE.test(engine)) throw new WireError("X_BAD_SEMVER", `engine version ${engine} is not semver`);

  if (status === "ERROR") {
    const err = r.error ?? "";
    if (!err) throw new WireError("X_EMPTY_VALUE", "ERROR status requires error= text");
    return `${PFR_PREFIX} | ${cid} | ERROR | engine=${encodeValue(engine)} ; error=${encodeValue(err)}`;
  }
  const findings = r.findings ?? [];
  if (!findings.length) throw new WireError("X_NO_FINDINGS", "a PFR carries at least one finding token");
  const tokens = [`engine=${encodeValue(engine)}`, ...findings.map(findingToken)];
  return `${PFR_PREFIX} | ${cid} | ${status} | ${tokens.join(" ; ")}`;
}

File: engine/wire.py
"""PFQ v1 / PFR v1 wire-format layer: pure parsing and rendering.

Frozen contract (spec/preflight-validator.md §3):

    Request:  PFQ v1 | <cid> | <op> | reply=<mb-room> ; k=v ; k=v …
    Response: PFR v1 | <cid> | <STATUS> | engine=<semver> ; findings…

This module adds ZERO entries to the frozen closed vocabularies (STATUS,
T1-reject, T1-warn, T2-observe sets live verbatim in REJECT_CODES /
WARN_CODES / OBSERVE_CODES below and must never gain members here).
Parse-layer failures use a SEPARATE X_* namespace so extension cannot be
confused with modifying the frozen sets.

Canonical spacing: single spaces around " | " and " ; ". Parsers accept
variable surrounding whitespace and normalize; renderers emit only the
canonical form, so render(parse(x)) is a fixed point and
parse(render(struct)) == struct whenever the struct is expressible
(spec-permitted round-trip).

Structural characters: "|" separates major fields, ";" separates parameter
tokens. Inside values/details they (and "%") MUST be percent-encoded --
only three escapes exist (%25 %7C %3B, uppercase), anything else is a
decoding error. Everything is decided by total functions of the input;
no clock, no randomness, no I/O.
"""
from __future__ import annotations

import re

from engine import preflight as pf
from engine.preflight import PreflightError

# ------------------------------------------------------------------ frozen
PFQ_PREFIX = "PFQ v1"
PFR_PREFIX = "PFR v1"
OPS = ("preview", "verify", "audit-did-note")
STATUSES = ("PASS", "FAIL", "PARTIAL", "ERROR")

# Frozen closed vocabularies -- DO NOT EDIT (spec §3). Parse errors use X_*.
REJECT_CODES = frozenset({
    "E_EMPTY_AFTER_SWEEP", "E_TEXT_TOO_LONG", "E_BAD_ROOM",
    "E_BAD_NONCE_FORMAT", "E_BAD_DID", "E_BAD_SIG_ENCODING",
    "E_SIG_INVALID", "E_CANONICAL_TOO_LONG",
})
WARN_CODES = frozenset({
    "W_SWEPT_CHARS", "W_URL_LONG", "W_LEADING_ZERO_NONCE",
    "W_NOTE_WRONG_KEY", "W_NOTE_LEGACY_PATH", "W_NOTE_FIELD_MISMATCH",
})
OBSERVE_CODES = frozenset({
    "O_NONCE_FLOOR_VISIBLE", "O_NO_PRIOR_WRITES_SEEN",
    "O_ROOM_OWNED", "O_CAPACITY_TIGHT",
})

# ------------------------------------------------------------- wire grammar
CID_RE = re.compile(r"[0-9a-f]{16}")
SEMVER_RE = re.compile(r"[0-9]+\.[0-9]+\.[0-9]+(?:-[0-9A-Za-z.-]+)?")
KEY_RE = re.compile(r"[a-z][a-z0-9_]{0,23}")
SHA256_RE = re.compile(r"[0-9a-f]{64}")
FP_RE = re.compile(r"[0-9a-f]{16}")
NS_RE = re.compile(r"did(?:-[0-9a-f]{2})?")
NOTE_KEY_RE = re.compile(r"[a-z0-9][a-z0-9_-]{0,47}")
REF_RE = re.compile(r"[0-9A-Za-z._:@/-]{1,40}")
MAX_LINE_CHARS = pf.MAX_TEXT_CHARS          # a request IS one stored message

ESCAPES = {"25": "%", "7C": "|", "3B": ";"}

# Op schemas straight from spec §2. reply is universal and handled separately.
OP_SCHEMAS: dict[str, tuple[tuple[str, ...], tuple[str, ...]]] = {
    #            required                          optional
    "preview": (("room", "nonce", "did", "text"), ("sig",)),
    "verify":  (("did",), ("sig", "nonce", "room", "text",
                           "canonical", "sha256")),
    "audit-did-note": (("value",), ("did", "fp", "ns", "key")),
}


class WireError(Exception):
    """Deterministic parse failure. .code is an X_* code, .detail explains."""

    def __init__(self, code: str, detail: str = ""):
        super().__init__(f"{code}: {detail}" if detail else code)
        self.code = code
        self.detail = detail


# ----------------------------------------------------------- value escaping
def encode_value(value: str) -> str:
    out = []
    for ch in value:
        if ch == "%":
            out.append("%25")
        elif ch == "|":
            out.append("%7C")
        elif ch == ";":
            out.append("%3B")
        else:
            out.append(ch)
    return "".join(out)


def decode_value(value: str) -> str:
    out: list[str] = []
    i = 0
    while i < len(value):
        ch = value[i]
        if ch != "%":
            out.append(ch)
            i += 1
            continue
        pair = value[i + 1:i + 3]
        if pair not in ESCAPES:
            raise WireError(
                "X_BAD_ENCODING",
                f"only %25 %7C %3B escapes exist; found %{pair!r} at offset {i}")
        out.append(ESCAPES[pair])
        i += 3
    return "".join(out)


# ------------------------------------------------------------ shared pieces
def _check_line(line: str) -> None:
    if not isinstance(line, str):
        raise WireError("X_BAD_TYPE", "wire line must be str")
    if len(line) > MAX_LINE_CHARS:
        raise WireError("X_LINE_TOO_LONG",
                        f"{len(line)} chars exceeds message cap {MAX_LINE_CHARS}")
    try:
        swept = pf.sweep(line)
    except PreflightError:
        raise WireError("X_EMPTY_AFTER_SWEEP", "line has no visible content")
    if swept.stored != line:
        raise WireError(
            "X_NOT_SWEEP_SAFE",
            "line contains characters the server would replace or strip; "
            "send the post-sweep form")


def _parse_cid(field: str) -> str:
    cid = field.strip()
    if not CID_RE.fullmatch(cid):
        raise WireError(
            "X_BAD_CID",
            f"cid must be exactly 16 lowercase hex chars, got {field!r}")
    return cid


def _split_params(raw: str) -> list[list[str]]:
    """Split the params section on ';' into (key, decoded_value) pairs,
    enforcing token shape. Returns ordered pairs."""
    pairs: list[list[str]] = []
    for tok in raw.split(";"):
        tok = tok.strip()
        if not tok:
            raise WireError("X_BAD_PARAM", "empty parameter token")
        eq = tok.find("=")
        if eq <= 0:
            raise WireError("X_BAD_PARAM",
                            f"parameter {tok!r} lacks 'key=value' form")
        key = tok[:eq].strip()
        if not KEY_RE.fullmatch(key):
            raise WireError("X_BAD_KEY",
                            f"parameter key {key!r} must match [a-z][a-z0-9_]{{0,23}}")
        val = tok[eq + 1:].strip()
        if not val:
            raise WireError("X_EMPTY_VALUE", f"parameter {key!r} has an empty value")
        pairs.append([key, decode_value(val)])
    return pairs


def _dup_check(pairs: list[list[str]]) -> dict[str, str]:
    seen: dict[str, str] = {}
    for key, val in pairs:
        if key in seen:
            raise WireError("X_DUPLICATE_KEY", f"parameter {key!r} appears twice")
        seen[key] = val
    return seen


def _require_first(pairs: list[list[str]], key: str) -> None:
    if not pairs or pairs[0][0] != key:
        raise WireError("X_ORDER", f"first parameter must be {key}=…")


def _validate_reply(room: str) -> None:
    try:
        v = pf.validate_room(room)
        if not v.ok:
            raise WireError("X_BAD_REPLY_ROOM", v.detail)
        if "mb" not in pf._room_classes(room):
            raise WireError("X_BAD_REPLY_ROOM",
                            f"{room!r} lacks the mb- class; replies go to mailboxes")
    except PreflightError as exc:
        raise WireError("X_BAD_REPLY_ROOM", str(exc)) from exc


def _finding_token(f: tuple) -> str:
    kind, code, ref, detail = f
    enc = encode_value(detail)
    if kind == "T1-ok":
        return "T1-ok" + (f" {enc}" if enc else "")
    if kind in ("T1-reject", "T1-warn"):
        return f"{kind}:{code}" + (f" {enc}" if enc else "")
    if kind == "T2-observe":
        base = f"T2-observe:{code}@{ref}"
        return base + (f" {enc}" if enc else "")
    raise WireError("X_BAD_FINDING", f"unrenderable finding kind {kind!r}")


# ------------------------------------------------------------------- PFQ
def parse_pfq(line: str) -> dict:
    _check_line(line)
    parts = line.split("|")
    if len(parts) != 4:
        raise WireError(
            "X_BAD_STRUCTURE",
            f"expected exactly 3 unescaped '|' separators, got {len(parts) - 1}")
    prefix, cid_f, op_f, params_f = (p.strip() for p in parts)
    if prefix != PFQ_PREFIX:
        raise WireError("X_BAD_PREFIX", f"expected {PFQ_PREFIX!r}, got {prefix!r}")
    cid = _parse_cid(cid_f)
    op = op_f.strip()
    if op not in OPS:
        raise WireError("X_BAD_OP",
                        f"op {op!r} outside {OPS}")
    pairs = _split_params(params_f)
    _require_first(pairs, "reply")
    kv = _dup_check(pairs)
    _validate_reply(kv["reply"])

    required, optional = OP_SCHEMAS[op]
    unknown = set(kv) - {"reply"} - set(required) - set(optional)
    if unknown:
        raise WireError("X_UNKNOWN_KEY",
                        f"op {op!r} does not accept {sorted(unknown)}")
    missing = [k for k in required if k not in kv]
    if missing:
        raise WireError("X_MISSING_KEY", f"op {op!r} requires {missing}")

    # closed-format parameters
    if "sha256" in kv and not SHA256_RE.fullmatch(kv["sha256"]):
        raise WireError("X_BAD_SHA256", "must be 64 lowercase hex chars")
    if "fp" in kv and not FP_RE.fullmatch(kv["fp"]):
        raise WireError("X_BAD_FP", "must be 16 lowercase hex chars")
    if "ns" in kv and not NS_RE.fullmatch(kv["ns"]):
        raise WireError("X_BAD_NS", "expected 'did' or 'did-<2 hex>'")
    if "key" in kv and not NOTE_KEY_RE.fullmatch(kv["key"]):
        raise WireError("X_BAD_NOTE_KEY", "invalid note key name")

    # op-specific completeness (spec §2B/§2C)
    if op == "verify":
        full = ("nonce" in kv, "room" in kv, "text" in kv)
        privacy = ("canonical" in kv, "sha256" in kv)
        if any(full) and not all(full):
            raise WireError("X_AMBIGUOUS_MODE",
                            "full mode needs nonce, room AND text together")
        if any(full) and any(privacy):
            raise WireError("X_AMBIGUOUS_MODE",
                            "full mode and privacy mode are mutually exclusive")
        if not any(full):
            if not any(privacy):
                raise WireError("X_MISSING_KEY",
                                "verify needs nonce+room+text or canonical+sha256")
            if not all(privacy):
                raise WireError("X_AMBIGUOUS_MODE",
                                "privacy mode needs canonical AND sha256 together")
    if op == "audit-did-note":
        has_did, has_fp = "did" in kv, "fp" in kv
        if has_did and has_fp:
            raise WireError("X_AMBIGUOUS_MODE",
                            "give exactly one of did= or fp=")
        if not has_did and not has_fp:
            raise WireError("X_MISSING_KEY",
                            "audit-did-note needs did= or fp=")

    return {"kind": "PFQ", "cid": cid, "op": op,
            "params": {k: v for k, v in kv.items()}}


def render_pfq(q: dict) -> str:
    """Render the canonical PFQ v1 line for a parsed/constructed struct."""
    op = q["op"]
    if op not in OPS:
        raise WireError("X_BAD_OP", f"op {op!r} outside {OPS}")
    cid = q.get("cid", "")
    if not CID_RE.fullmatch(cid):
        raise WireError("X_BAD_CID", "cid must be 16 lowercase hex chars")
    kv = q.get("params", {})
    _dup_render_check(kv)
    if "reply" not in kv:
        raise WireError("X_MISSING_KEY", "reply= is mandatory")
    required, optional = OP_SCHEMAS[op]
    unknown = set(kv) - {"reply"} - set(required) - set(optional)
    if unknown:
        raise WireError("X_UNKNOWN_KEY",
                        f"op {op!r} does not accept {sorted(unknown)}")
    missing = [k for k in required if k not in kv]
    if missing:
        raise WireError("X_MISSING_KEY", f"op {op!r} requires {missing}")
    ordered = ["reply"] + [k for k in kv if k != "reply"]
    params = " ; ".join(f"{k}={encode_value(kv[k])}" for k in ordered)
    return f"{PFQ_PREFIX} | {cid} | {op} | {params}"


def _dup_render_check(kv: dict) -> None:
    if len(kv) != len(set(kv)):
        raise WireError("X_DUPLICATE_KEY", "dict cannot hold duplicate keys")


# ------------------------------------------------------------------- PFR
def parse_pfr(line: str) -> dict:
    _check_line(line)
    parts = line.split("|")
    if len(parts) != 4:
        raise WireError(
            "X_BAD_STRUCTURE",
            f"expected exactly 3 unescaped '|' separators, got {len(parts) - 1}")
    prefix, cid_f, status_f, params_f = (p.strip() for p in parts)
    if prefix != PFR_PREFIX:
        raise WireError("X_BAD_PREFIX", f"expected {PFR_PREFIX!r}, got {prefix!r}")
    cid = _parse_cid(cid_f)
    status = status_f.strip()
    if status not in STATUSES:
        raise WireError("X_BAD_STATUS",
                        f"status {status!r} outside {STATUSES}")

    tokens = [t.strip() for t in params_f.split(";")]
    if not tokens or not tokens[0]:
        raise WireError("X_BAD_PARAM", "empty parameter section")
    # First token is the only key=value: engine=<semver>. Findings are bare
    # tokens whose details carry spaces, so they must NOT go through k=v logic.
    eq = tokens[0].find("=")
    if eq <= 0 or tokens[0][:eq].strip() != "engine":
        raise WireError("X_ORDER", "first parameter must be engine=<semver>")
    engine = decode_value(tokens[0][eq + 1:].strip())
    if not engine:
        raise WireError("X_EMPTY_VALUE", "engine has an empty value")
    if not SEMVER_RE.fullmatch(engine):
        raise WireError("X_BAD_SEMVER",
                        f"engine version {engine!r} is not semver")

    # ERROR responses carry error=<text> instead of finding tokens
    # (transport-layer completion rule mirrored from render_pfr).
    if status == "ERROR":
        if len(tokens) != 2:
            raise WireError("X_BAD_STRUCTURE",
                            "ERROR response carries exactly one error= token")
        eq2 = tokens[1].find("=")
        if eq2 <= 0 or tokens[1][:eq2].strip() != "error":
            raise WireError("X_BAD_PARAM",
                            "ERROR response requires error=<text>")
        err_text = decode_value(tokens[1][eq2 + 1:].strip())
        if not err_text:
            raise WireError("X_EMPTY_VALUE", "error= text is empty")
        return {"kind": "PFR", "cid": cid, "status": status,
                "engine": engine, "findings": [], "error": err_text}

    findings = []
    for tok in tokens[1:]:
        if not tok:
            raise WireError("X_BAD_PARAM", "empty parameter token")
        head, _, detail = tok.partition(" ")
        detail = decode_value(detail.strip())
        if head == "T1-ok":
            findings.append(("T1-ok", "", "", detail))
            continue
        m = re.fullmatch(r"(T1-reject|T1-warn):([A-Z0-9_]+)", head)
        if m:
            kind, code = m.groups()
            table = REJECT_CODES if kind == "T1-reject" else WARN_CODES
            if code not in table:
                raise WireError(
                    "X_UNKNOWN_FINDING_CODE",
                    f"{kind}:{code} is outside the frozen {kind} vocabulary")
            findings.append((kind, code, "", detail))
            continue
        m = re.fullmatch(r"T2-observe:([A-Z0-9_]+)@(.*)", head)
        if m:
            code, ref = m.groups()
            if code not in OBSERVE_CODES:
                raise WireError(
                    "X_UNKNOWN_FINDING_CODE",
                    f"T2-observe:{code} is outside the frozen T2 vocabulary")
            if not ref or not REF_RE.fullmatch(ref):
                raise WireError("X_BAD_OBSERVATION_REF",
                                f"observation ref {ref!r} is malformed")
            findings.append(("T2-observe", code, ref, detail))
            continue
        raise WireError("X_BAD_FINDING", f"finding token {head!r} is malformed")

    if not findings:
        raise WireError("X_NO_FINDINGS",
                        "a PFR carries at least one finding token")
    return {"kind": "PFR", "cid": cid, "status": status,
            "engine": engine, "findings": findings}


def render_pfr(r: dict) -> str:
    status = r.get("status")
    if status not in STATUSES:
        raise WireError("X_BAD_STATUS", f"status {status!r} outside {STATUSES}")
    cid = r.get("cid", "")
    if not CID_RE.fullmatch(cid):
        raise WireError("X_BAD_CID", "cid must be 16 lowercase hex chars")
    engine = r.get("engine", "")
    if not SEMVER_RE.fullmatch(engine):
        raise WireError("X_BAD_SEMVER", f"engine version {engine!r} is not semver")

    if status == "ERROR":
        # Transport-layer completion (2026-08-26): ERROR responses carry
        # error=<encoded text> instead of finding tokens. Additive parameter;
        # frozen finding vocabularies untouched.
        err = r.get("error", "")
        if not err:
            raise WireError("X_EMPTY_VALUE", "ERROR status requires error= text")
        return (f"{PFR_PREFIX} | {cid} | ERROR | "
                f"engine={encode_value(engine)} ; error={encode_value(err)}")

    findings = r.get("findings") or ()
    if not findings:
        raise WireError("X_NO_FINDINGS",
                        "a PFR carries at least one finding token")
    tokens = [f"engine={encode_value(engine)}"] + \
             [_finding_token(f) for f in findings]
    return f"{PFR_PREFIX} | {cid} | {status} | {' ; '.join(tokens)}"
